facade/src/main/java/org/jboss/pnc/facade/providers/BuildFetcher.java

```diff
@@ -0,0 +1,426 @@
+    private Map<String, BuildTask> buildTaskCache = new HashMap<>();
+    private Map<Base32LongID, BuildRecord> buildRecordCache = new HashMap<>();
+    private Set<Base32LongID> dependenciesToFetch = new HashSet<>();
+
+    public BuildWithDependencies getBuildWithDependencies(String id) {
+        BuildTask buildTask = buildTaskCache.get(id);
+        if (buildTask != null) {
+            return new BuildWithDependencies(buildMapper.fromBuildTask(buildTask), buildTask);
+        }
+        Base32LongID dbid = buildMapper.getIdMapper().toEntity(id);
+        BuildRecord buildRecord = buildRecordCache.get(dbid);
+        if (buildRecord == null) {
+            buildRecord = buildRecordRepository.queryById(dbid);
+            if (buildRecord == null) {
+                throw new CorruptedDataException("Missing build with id:" + id);
+            }
+            log.warn("Prefetch missed build " + id);
+            buildRecordCache.put(buildRecord.getId(), buildRecord);
+        }
+        return new BuildWithDependencies(buildMapper.toDTO(buildRecord), buildRecord);
+    }
+
+    public static Comparator<BuildWrapper> wrapperComparator(Comparator<Build> comparator) {
+        return (o1, o2) -> (comparator.compare(o1.mappedBuild, o2.mappedBuild));
+    }
+
+    public List<Build> getBuildPage(
+            int pageIndex,
+            int pageSize,
+            List<Build> runningBuilds,
+            Predicate<BuildRecord>[] predicates,
+            SortInfo<BuildRecord> sortInfo,
+            Comparator<Build> comparing) {
+
+        int firstPossibleDBIndex = pageIndex * pageSize - runningBuilds.size();
+        int lastPossibleDBIndex = (pageIndex + 1) * pageSize - 1;
+        int toSkip = min(runningBuilds.size(), pageIndex * pageSize);
+
+        Iterator<BuildWrapper> wrappedRunningBuilds = runningBuilds.stream()
+                .sorted(comparing)
+                .map(BuildWrapper::new)
+                .iterator();
+        BuildIterator buildIterator = new BuildIterator(
+                firstPossibleDBIndex,
+                lastPossibleDBIndex,
+                pageSize,
+                sortInfo,
+                predicates);
+        MergeIterator<BuildWrapper> buildsIT = new MergeIterator<>(
+                wrappedRunningBuilds,
+                buildIterator,
+                wrapperComparator(comparing));
+        List<BuildWrapper> builds = StreamSupport
+                .stream(Spliterators.spliteratorUnknownSize(buildsIT, Spliterator.ORDERED | Spliterator.SORTED), false)
+                .skip(toSkip)
+                .limit(pageSize)
+                .collect(Collectors.toList());
+
+        fetchBuildConfigAudited(builds.stream().flatMap(BuildWrapper::buildRecordStream).collect(Collectors.toSet()));
+        return builds.stream().map(BuildWrapper::getBuild).collect(Collectors.toList());
+    }
+
+    /**
+     * Fetch all running builds from build coordinator and cache them.
+     */
+    public void fetchRunningBuilds() {
```

Comment:
private?

Reply:
public, it's used in tests and could theoretically be used by the BuildFetcher users as well